README.md
# strong-statsd

Control a statsd child process.

Uses strongloop-forks/strong-fork-statsd, a fork of etsy/statsd with support
for being controlled by a parent process.

## Usage

### Statsd = require('strong-statsd')

Constructor.

### statsd = Statsd(options)

Options:

- debug: cause statsd to log debug messages
- scope: prefix to add before all metric names (default is `""`, but also,
  if the `statsd:` backend URL is used, it can provide a scope that will
  override this option)
- expandScope: function to call on scope before using it as a prefix
- flushInterval: accumulated metrics will be processed and emitted at this
  interval, in seconds (default is 15)

### statsd.backend(url)

Specify one or more backends.

Backend URL formats:

- `statsd://[<host>][:<port>]`: publish metrics to a statsd receiver. The host
  defaults to `"localhost"`, and the port defaults to `8125`.

- `log:[<file>]`: log metrics to FILE, which may be `-` to indicate stdout (FILE
  defaults to `-`). Format is `ISOTIMESTAMP METRIC=VALUE (TYPE)`, where TYPE is
  one of count, gauge, or ms (a timer interval in milliseconds).

- `debug:[?pretty[=<true|false>]]`: json dump to stdout for testing and
  debugging backends. Pretty output is formatted as multi-line with color,
  otherwise it's single line.

- `graphite://[<host>][:<port>]`: forward to
  [graphite](http://graphite.readthedocs.org/en/latest/), host defaults to
  `"localhost"`, port defaults to `2003`

- `syslog:[?[application=<application>][,priority=<priority>]`: write to
  local system log using `syslog(3)`. The application defaults to `"statsd"`,
  and priority defaults to `"LOG_INFO"`, but can be set to any of `"LOG_DEBUG"`,
  `"LOG_INFO"`, `"LOG_NOTICE"`, `"LOG_WARNING"`, or `"LOG_CRIT"`.

- `splunk://[<host>]:<port>`: write using a UDP key value protocol to splunk,
  host defaults to localhost, and port is mandatory, since the protocol has no
  assigned port.

Throws on error in URL format, returns self on success.

### statsd.start(callback)

Starts statsd child process, callback indicates when it is started, and the
statsd port is known.

### statsd.port

The port that the statsd server is listening on.

### statsd.url

A `statsd:` URL that contains the PORT and (unexpanded) scope that should be
published to. The publish is usually done by `.send()`, but this URL may
be useful for display or debug purposes.

### statsd.send(name, value)

Send statsd metric with name and value. Metrics are discarded until statsd
is started.

### statsd.stop(callback)

Stop the statsd child, callback indicates it has exited.

Stopping is optional, the child will self-exit when the parent process exits.


CHANGES.md
2016-05-04, Version 2.0.6
=========================

 * package: do not strip test servers (Sam Roberts)


2016-04-11, Version 2.0.5
=========================

 * Refer to licenses with a link (Sam Roberts)


2015-10-05, Version 2.0.4
=========================

 * Use strongloop conventions for licensing (Sam Roberts)


2015-09-16, Version 2.0.3
=========================

 * syslog: replace node-syslog with modern-syslog (Sam Roberts)


2015-02-11, Version 2.0.2
=========================

 * Fix timestamps in log backend (Ryan Graham)


2015-01-12, Version 2.0.1
=========================

 * Fix bad CLA URL in CONTRIBUTING.md (Ryan Graham)


2014-12-15, Version 2.0.0
=========================

 * fix syslog test plan (Sam Roberts)

 * Use cluster IPC, not UDP, to send stats to master (Sam Roberts)


2014-12-05, Version 1.1.1
=========================

 * package: fix syslog docs (`,` should be `&`) (Sam Roberts)

 * test: replace tap fork with upstream (Ryan Graham)

 * syslog: replace node-syslog with strong-fork-syslog (Ryan Graham)


2014-12-01, Version 1.1.0
=========================

 * Use relative backend names, the logs are cleaner (Sam Roberts)

 * test: server APIs made more uniform (Sam Roberts)

 * test: syslog server has addon dependency injected (Sam Roberts)

 * syslog: add debug logging of messages (Sam Roberts)

 * Error if no backends, or multiple backends. (Sam Roberts)

 * test: add unref() to graphite server (Sam Roberts)

 * internal: use javascript milliseconds since epoc (Sam Roberts)

 * syslog: node-syslog dependency is now injected (Sam Roberts)

 * syslog: reimplement as a backend (Sam Roberts)

 * Add copyright header to lib/log.js (Sam Roberts)

 * internal-statsd: reimplement worker/master statsd (Sam Roberts)

 * test: refactor into reuseable test servers (Sam Roberts)

 * Fix incomplete rewrite during rebase (Sam Roberts)

 * statsd: use the repeater to forward metrics (Sam Roberts)

 * splunk: support checking of splunk host name (Sam Roberts)

 * Support async checking of backends (Sam Roberts)

 * Direct use of statsd, not through child process (Sam Roberts)

 * Import core etsty/statsd code (Sam Roberts)

 * Add strongloop copyright headers (Sam Roberts)


2014-11-14, Version 1.0.0
=========================

 * package: mark version as -initial (Sam Roberts)

 * internal: work around UPD message ordering (Sam Roberts)

 * test: add test for negative gauge values (Sam Roberts)

 * statsd: support a configurable flush interval (Sam Roberts)

 * internal: emit metrics internally from the backend (Sam Roberts)

 * package: node-syslog dependency is optional (Ryan Graham)

 * statsd: set server url on start (Sam Roberts)

 * test: only report PASS if exit code is OK (Sam Roberts)

 * Support log: URL, to log metrics to file or stdout (Sam Roberts)

 * Rename console backend to debug (Sam Roberts)

 * Throw on error specifying the backend. (Sam Roberts)

 * package: increment version to 1.0.0 for staging (Sam Roberts)

 * Keep child refed until it has started (Sam Roberts)

 * Console color option now controls prettyprinting (Sam Roberts)

 * Support statsd:// URL, and sending metrics (Sam Roberts)

 * test: check that statsd is unrefed (Sam Roberts)

 * Start a statsd daemon, configured via URL (Sam Roberts)

 * Initial package (Sam Roberts)


2014-10-03, Version INITIAL
===========================

 * First release!


index.js
// Copyright (C) 2014 Strongloop, see LICENSE.md
var EventEmitter = require('events').EventEmitter;
var Log = require('./lib/log');
var Server = require('./lib/server');
var assert = require('assert');
var debug = require('debug')('strong-statsd');
var fmt = require('util').format;
var fork = require('child_process').fork;
var fs = require('fs');
var parse = require('url').parse;
var path = require('path');
var sender = require('strong-agent-statsd');
var util = require('util');

// Config template:
// {
//   // Start listening for statsd/udp on ephemeral port
//   port: 0,
// 
//   // Graphite configuration
//   backends: [ "./backends/graphite" ],
//   graphitePort: 2003,
//   graphiteHost: "localhost",
//   graphite: { legacyNamespace: false, },
// 
//   // Syslog configuration (should be a backend, but it isn't)
//   dumpMessages: true,
//   log: { backend: 'syslog', level: 'LOG_WARNING' },
// 
//   // Console configuration, useful for testing
//   backends: [ "./backends/console" ],
//   console: { prettyprint: true },
//
//   // splunk configuration:
//   //  https://github.com/dylanmei/statsd-udpkv-backend#configuration
// }

function Statsd(options) {
  if (!(this instanceof Statsd))
      return new Statsd(options);

  EventEmitter.call(this);

  options = util._extend({}, options);
  this.port = 0;
  // For a worker we should directly send using statsd to the master using
  // the internal port.
  this.internalPort = null;
  this.debug = !!options.debug;
  this.expandScope = options.expandScope;
  this.scope = options.scope || '';
  this.configFile = path.resolve('.statsd.json');
  this.flushInterval = (options.flushInterval || 15) * 1000;
  this.syslog = options.syslog; // node-syslog dependency must be provided
  this.config = {
    port: this.port,
    debug: this.debug,
    flushInterval: this.flushInterval,
    dumpMessages: this.debug,
    backends: [],
  };
  this._send = null;
  this.server = null;
  this._socket = null;

  // XXX child.stdout/err structure for bacwards compat
  this.child = {
    stdout: new Log('stdout'),
    stderr: new Log('stderr'), // FIXME unused?
  };
  this.logger = {
    log: this.child.stdout.log.bind(this.child.stdout),
  };

}

util.inherits(Statsd, EventEmitter);

// XXX(sam) would be better to return a fully expanded URL (with all the
// defaults written in) then to return self
Statsd.prototype.backend = function backend(url) {
  var backend;
  var config = {};
  var _ = parse(url, true);
  if (!_.protocol) {
    // bare word, such as 'console', or 'statsd'
    _.protocol = url + ':';
    // bare word also shows up in path & pathname, clear it
    _.pathname = '';
    _.path = '';
  }

  switch (_.protocol) {
    case 'statsd:': {
      if (_.path && _.path !== '/')
        die('statsd scope not supported');
      backend = "./backends/repeater";
      config = {
        repeater: [{
          host: _.hostname || 'localhost',
          port: _.port || 8125,
        }]
      };
      break;
    }
    case 'debug:': {
      backend = "./backends/console";
      config = {
        console: {
          prettyprint: 'pretty' in _.query && _.query.pretty !== 'false'
        }
      };
      break;
    }
    case 'log:': {
      backend = './backends/log';
      config = {
        log: {
          file: (_.hostname || '') + (_.pathname || ''),
          stdout: this.child.stdout,
        }
      };
      config.log.file = config.log.file || '-';
      break;
    }
    case 'graphite:': {
      backend = "./backends/graphite";
      config = {
        graphitePort: _.port || 2003, // graphite default
        graphiteHost: _.hostname || 'localhost',
        graphite: {
          legacyNamespace: false
        }
      };
      break;
    }
    case 'splunk:': {
      if (!_.port) {
        return die('splunk port missing');
      }
      backend = "./backends/splunk";
      config = {
        udpkv: {
          host: _.hostname || 'localhost',
          port: _.port,
        },
      };
      break;
    }
    case 'syslog:': {
      if (!this.syslog) {
        return die('syslog not supported');
      }
      var priority = _.query.priority || 'LOG_INFO';
      if (priority) {
        // Must be valid, or syslog will abort.
        if (!/^LOG_/.test(priority) || !(priority in this.syslog)) {
          return die('syslog priority invalid');
        }
      }
      backend = './backends/syslog';
      config = {
        syslog: {
          application: _.query.application || 'statsd',
          priority: priority,
          syslog: this.syslog,
        }
      };
      break;
    }
    case 'internal:': {
      backend = './backends/internal';
      config = {
        internal: {
          notify: this.emit.bind(this, 'metrics'),
        },
      };
      break;
    }
    case 'internal-statsd:': {
      assert(_.port && _.port > 0, 'malformed internal-statsd URL: '+url);
      this.internalPort = _.port;
      break;
    }
    default:
      return die('url format unknown');
  }

  if (this.internalPort && this.config.backends.length) {
    die('misconfigured, cannot be internal and have a backend');
  }

  if (backend) {
    if (this.config.backends.indexOf(backend) > -1) {
      throw Error(fmt('%s metrics already configured', _.protocol));
    }
    this.config.backends.push(backend);
  }

  this.config = util._extend(this.config, config);

  function die(error) {
    var er = Error(error);
    er.url = url;
    throw er;
  }

  return this;
};

Statsd.prototype.start = function start(callback) {
  var self = this;
  var scope = this.scope;
  scope = this.expandScope ? this.expandScope(scope) : scope;

  if (this.internalPort) {
    // Directly send to the internal/master statsd port
    self._send = sender({
      port: this.internalPort,
      host: 'localhost',
      scope: scope,
    });

    self.url = fmt('internal-statsd://:%d', self.internalPort);

    process.nextTick(callback);
    return;
  }

  debug('statsd config: %j', this.config);

  if (this.config.backends.length < 1) {
    process.nextTick(function() {
      callback(Error('start failed: no backends configured'));
    });
    return;
  }

  this.server = new Server;
  this.server.start(this.config, this.logger.log, onStart);

  function onStart(er, server) {
    if (er) return callback(er);

    self._socket = server;
    self.port = server.address().port;

    self._send = sender({
      port: self.port,
      host: 'localhost',
      scope: scope,
    });

    self.url = fmt('internal-statsd://:%d', self.port);

    callback();
  }

  return this;
};

Statsd.prototype.send = function send(name, value) {
  if (this._send) {
    this._send(name, value);
    return true;
  }
  return false;
};

Statsd.prototype.stop = function stop(callback) {
  this.stopped = true;

  callback = callback || function(){};
  if (!this.server) {
    process.nextTick(callback);
    return;
  }
  this.server.stop(callback);
  this.server = null;
};

module.exports = Statsd;
